File: sdk-java/worker/src/main/java/io/github/srjn45/rdq/worker/engine/Backoff.java
/*
 * Copyright 2025-2026 Srajan Pathak
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.srjn45.rdq.worker.engine;

import java.time.Duration;
import java.util.Objects;
import java.util.function.DoubleSupplier;

/**
 * The resolved retry-backoff ladder for a queue (design 03 &sect;2). Mirrors Go
 * {@code core/policy.Backoff}: concrete values, decoupled from config pointers, so
 * {@link #delay} is a pure numeric function.
 *
 * <p>Formula: {@code delay(n) = min(initial × multiplier^(n−1), max) × (1 ± jitter·rand)}
 * where the cap is applied to the exponential term <em>before</em> jitter, exactly
 * as the Go reference implementation.
 *
 * <p>Thread-safe (immutable).
 */
public final class Backoff {

    private final long initialNanos;
    private final double multiplier;
    private final long maxNanos;
    private final double jitter;

    private Backoff(Builder b) {
        this.initialNanos = b.initial.toNanos();
        this.multiplier = b.multiplier;
        this.maxNanos = b.max.toNanos();
        this.jitter = b.jitter;
    }

    /**
     * Computes the wait before retry attempt {@code n} (1-based). When
     * {@code jitter == 0}, {@code rng} is never consulted and may be {@code null}.
     */
    public Duration delay(int n, DoubleSupplier rng) {
        long base = baseDelayNanos(n);
        if (jitter <= 0) {
            return Duration.ofNanos(base);
        }
        double factor = 1.0 + jitter * (2.0 * rng.getAsDouble() - 1.0);
        long d = (long) ((double) base * factor);
        if (d < 0) return Duration.ZERO;
        return Duration.ofNanos(d);
    }

    private long baseDelayNanos(int n) {
        if (n < 1) n = 1;
        double scaled = (double) initialNanos * Math.pow(multiplier, n - 1);
        if (maxNanos > 0 && scaled >= (double) maxNanos) {
            return maxNanos;
        }
        if (scaled >= (double) Long.MAX_VALUE) {
            return Long.MAX_VALUE;
        }
        return (long) scaled;
    }

    public static Builder builder() {
        return new Builder();
    }

    /** Fluent builder for {@link Backoff}. */
    public static final class Builder {

        private Duration initial = Duration.ofSeconds(1);
        private double multiplier = 1.0;
        private Duration max = Duration.ZERO; // 0 = uncapped
        private double jitter = 0.0;

        private Builder() {}

        /** Base delay before the first retry (n=1). Default: 1&nbsp;s. */
        public Builder initial(Duration initial) {
            this.initial = Objects.requireNonNull(initial, "initial");
            return this;
        }

        /**
         * Multiplier applied per attempt; 1.0 is linear (constant). Default: 1.0.
         * Config validation requires &ge;&nbsp;1.0.
         */
        public Builder multiplier(double multiplier) {
            this.multiplier = multiplier;
            return this;
        }

        /**
         * Cap on the pre-jitter delay. Zero (default) means uncapped.
         */
        public Builder max(Duration max) {
            this.max = Objects.requireNonNull(max, "max");
            return this;
        }

        /**
         * Symmetric spread as a fraction of the delay in [0, 1]: the result lands
         * in {@code base·[1−jitter, 1+jitter)}. Zero (default) disables jitter.
         */
        public Builder jitter(double jitter) {
            this.jitter = jitter;
            return this;
        }

        public Backoff build() {
            return new Backoff(this);
        }
    }
}

File: core/policy/backoff.go
// SPDX-License-Identifier: Apache-2.0

// Package policy holds the engine's pure decision functions — the retry
// backoff ladder here, outcome classification alongside it (design 03 §2/§4).
// Everything in this package is deterministic given its inputs: the backoff
// takes an injected random source rather than reaching for a global RNG, so the
// engine's retry timing is fully reproducible under test (design 05, "backoff
// and classification are deterministic pure functions with heavy unit
// coverage").
package policy

import (
	"math"
	"time"

	"github.com/srjn45/rdq/core/config"
)

// RNG is the injected source of randomness for backoff jitter. It is satisfied
// by *math/rand.Rand and *math/rand/v2.Rand out of the box, and by a stub in
// tests — the whole point of injecting it is that a test can pin Float64 and
// assert an exact delay (design 03 §2, backlog T3.2).
//
// Float64 must return a value in [0, 1), the contract math/rand already
// guarantees.
type RNG interface {
	Float64() float64
}

// Backoff is the resolved retry-backoff ladder for a queue: concrete values,
// already merged and validated by the config loader (design 03 §2/§3). It is
// deliberately decoupled from *config.RetryConfig's inherit-or-set pointers so
// that Delay stays a pure numeric function; BackoffFromConfig bridges the two.
type Backoff struct {
	// Initial is the base delay before the first retry (n = 1).
	Initial time.Duration
	// Multiplier grows the delay per attempt; 1.0 is linear (constant). Config
	// validation guarantees it is ≥ 1.0.
	Multiplier float64
	// Max caps the pre-jitter delay. A non-positive Max means uncapped.
	Max time.Duration
	// Jitter is the symmetric spread as a fraction of the delay, in [0, 1]: the
	// jittered result lands in base·[1−Jitter, 1+Jitter). Zero disables jitter,
	// making Delay independent of the RNG.
	Jitter float64
}

// Delay computes the wait before retry attempt n (1-based), implementing the
// design's formula (design 03 §2):
//
//	delay(n) = min(initial × multiplier^(n−1), max) × (1 ± jitter·rand)
//
// The cap is applied to the exponential term before jitter, exactly as written,
// so a jittered delay may exceed max by up to the jitter fraction. The
// exponential is evaluated in float64 so a large n saturates to max instead of
// overflowing int64. When Jitter is zero, rng is never consulted and may be nil.
func (b Backoff) Delay(n int, rng RNG) time.Duration {
	base := b.baseDelay(n)
	if b.Jitter <= 0 {
		return base
	}
	// rng.Float64() ∈ [0, 1) maps to a factor in [1−Jitter, 1+Jitter).
	factor := 1 + b.Jitter*(2*rng.Float64()-1)
	d := time.Duration(float64(base) * factor)
	if d < 0 {
		// Only reachable with jitter > 1, which config validation forbids; guard
		// anyway so the engine never schedules a retry in the past.
		return 0
	}
	return d
}

// baseDelay is the pre-jitter min(initial × multiplier^(n−1), max). Attempt
// numbers below 1 are clamped to 1 so the exponent never goes negative.
func (b Backoff) baseDelay(n int) time.Duration {
	if n < 1 {
		n = 1
	}
	scaled := float64(b.Initial) * math.Pow(b.Multiplier, float64(n-1))
	if b.Max > 0 && scaled >= float64(b.Max) {
		return b.Max
	}
	if scaled >= float64(math.MaxInt64) {
		// Uncapped and overflowing (e.g. multiplier^n → +Inf); saturate rather
		// than wrap to a negative duration.
		return time.Duration(math.MaxInt64)
	}
	return time.Duration(scaled)
}

// BackoffFromConfig resolves a RetryConfig's backoff fields into a Backoff. It
// requires the four backoff fields to be present: the config loader's
// deep-merge fills them from defaults, and a queue whose effective retry block
// omits one has no defined ladder, so the engine — not this function — must
// decide that policy. ok is false when any required field is nil.
func BackoffFromConfig(rc *config.RetryConfig) (b Backoff, ok bool) {
	if rc == nil ||
		rc.InitialBackoff == nil ||
		rc.BackoffMultiplier == nil ||
		rc.MaxBackoff == nil ||
		rc.Jitter == nil {
		return Backoff{}, false
	}
	return Backoff{
		Initial:    rc.InitialBackoff.Std(),
		Multiplier: *rc.BackoffMultiplier,
		Max:        rc.MaxBackoff.Std(),
		Jitter:     *rc.Jitter,
	}, true
}
